### ISSUE TITLE
CTCLoss documentation is unclear

### ISSUE THREAD

**Issue by rlorigro** [pytorch/pytorch#18249]
## 📚 Documentation

Another student and I have been working out how to use the new CTCLoss function, and the documentation is somewhat incomplete. I want to point out some possible problems, and suggest changes to the example script.

Problems:
  - **S** is not defined for the shape of `targets`. S should be the length of the targets?
  - It appears that each value in **S** should be the index of the target, similar to CrossEntropy loss, but this is not explained
  - For the example code, why would the target indices start at 1 instead of 0?
  - `sum(target_lengths)` is not clear
  - For the example code, variables are not named in terms of **S,N,T,C**
  - It is unclear why `target_lengths` would ever need to contain values less than the true length of the target sequences, and it appears that "length" is actually the stop index? (in the example) I assume there is some intended reason for this, but it is a bit confusing why lengths are specified separately from the sequences themselves.

A possible improvement of the example script might look something like this (`x` and `y` could be replaced with `input` and `target`)

```
from torch import nn
import torch


T = 50  # Input sequence length
C = 20  # Number of classes
N = 16  # Batch size
S = 30  # Target sequence length

minimum_target_length = 10

ctc_loss = nn.CTCLoss()
x = torch.randn(T, N, C).log_softmax(2).detach().requires_grad_()    # [size] = T,N,C
y = torch.randint(1, C, (N, S), dtype=torch.long)                    # low, high, [size]

x_lengths = torch.full((N,), T, dtype=torch.long)                           # Length of inputs
y_lengths = torch.randint(minimum_target_length,S,(N,), dtype=torch.long)   # Length of targets can be variable (even if target sequences are constant length)

loss = ctc_loss(x, y, x_lengths, y_lengths)
loss.backward()

print("x.shape", x.shape)
print("y.shape", y.shape)
print("x_lengths", x_lengths)
print("y_lengths", y_lengths)

print(loss)
```

Hopefully this would save people some time figuring out the currently unstated requirements of CTCLoss.

credit to @Jacob0G for helping find some of these issues

**Comment by ezyang** [comment]
I'm happy to help review a doc PR, as many of us would be.

**Comment by rlorigro** [comment]
It would be helpful if someone could confirm that the things I have already suggested are actually correct. I could rewrite in terms of a PR, if that would also lead to an update on your website:

https://pytorch.org/docs/stable/nn.html#ctcloss

**Comment by t-vi** [comment]
@rlorigro This is highly appreciated, as would your PR! I'm sure the documentation can be much improved over what I put in there. (Maybe one could consider linking to the [distill.pub article](https://distill.pub/2017/ctc/) on CTC, too - but I haven't looked at the docs lately to see this.)
Some of your points are not yet actionable (e.g. that the targets start at one because 0 is blank in the example is written as a question).
I'd be very happy to review your PR, too, so please don't hesitate to tag me on it.

**Comment by rlorigro** [comment]
OK I made a PR for this. I couldn't find any documentation on how these are usually done, so I assume the idea is just to edit the docstring in `nn/modules/loss.py`
